File: clients/scala/src/main/scala/model/PlayerMessageGame.scala
package model

import util.StreamUtil

sealed trait PlayerMessageGame {
    def writeTo(stream: java.io.OutputStream)
}
object PlayerMessageGame {
    case class CustomDataMessage(data: model.CustomData) extends PlayerMessageGame {
        override def writeTo(stream: java.io.OutputStream) {
            StreamUtil.writeInt(stream, CustomDataMessage.TAG)
            data.writeTo(stream)
        }
    }
    object CustomDataMessage {
        val TAG: Int = 0
        def readFrom(stream: java.io.InputStream): CustomDataMessage = CustomDataMessage(
            model.CustomData.readFrom(stream)
            )
    }

    case class ActionMessage(action: model.Versioned) extends PlayerMessageGame {
        override def writeTo(stream: java.io.OutputStream) {
            StreamUtil.writeInt(stream, ActionMessage.TAG)
            action.writeTo(stream)
        }
    }
    object ActionMessage {
        val TAG: Int = 1
        def readFrom(stream: java.io.InputStream): ActionMessage = ActionMessage(
            model.Versioned.readFrom(stream)
            )
    }

    def readFrom(stream: java.io.InputStream): PlayerMessageGame = {
        StreamUtil.readInt(stream) match {
            case CustomDataMessage.TAG => CustomDataMessage.readFrom(stream)
            case ActionMessage.TAG => ActionMessage.readFrom(stream)
            case _ => throw new java.io.IOException("Unexpected discriminant value")
        }
    }
}

File: clients/java/src/main/java/model/PlayerMessageGame.java
package model;

import util.StreamUtil;

public abstract class PlayerMessageGame {
    public abstract void writeTo(java.io.OutputStream stream) throws java.io.IOException;
    public static PlayerMessageGame readFrom(java.io.InputStream stream) throws java.io.IOException {
        switch (StreamUtil.readInt(stream)) {
            case CustomDataMessage.TAG:
                return CustomDataMessage.readFrom(stream);
            case ActionMessage.TAG:
                return ActionMessage.readFrom(stream);
            default:
                throw new java.io.IOException("Unexpected discriminant value");
        }
    }

    public static class CustomDataMessage extends PlayerMessageGame {
        public static final int TAG = 0;
        private model.CustomData data;
        public model.CustomData getData() { return data; }
        public void setData(model.CustomData data) { this.data = data; }
        public CustomDataMessage() {}
        public CustomDataMessage(model.CustomData data) {
            this.data = data;
        }
        public static CustomDataMessage readFrom(java.io.InputStream stream) throws java.io.IOException {
            CustomDataMessage result = new CustomDataMessage();
            result.data = model.CustomData.readFrom(stream);
            return result;
        }
        @Override
        public void writeTo(java.io.OutputStream stream) throws java.io.IOException {
            StreamUtil.writeInt(stream, TAG);
            data.writeTo(stream);
        }
    }

    public static class ActionMessage extends PlayerMessageGame {
        public static final int TAG = 1;
        private model.Versioned action;
        public model.Versioned getAction() { return action; }
        public void setAction(model.Versioned action) { this.action = action; }
        public ActionMessage() {}
        public ActionMessage(model.Versioned action) {
            this.action = action;
        }
        public static ActionMessage readFrom(java.io.InputStream stream) throws java.io.IOException {
            ActionMessage result = new ActionMessage();
            result.action = model.Versioned.readFrom(stream);
            return result;
        }
        @Override
        public void writeTo(java.io.OutputStream stream) throws java.io.IOException {
            StreamUtil.writeInt(stream, TAG);
            action.writeTo(stream);
        }
    }
}
